File: tests/irregular/irregular.ml
module CompleteBinaryTree = struct
  type 'a t =
    | Leaf of 'a
    | Node of ('a * 'a) t

  let rec make : 'a . int -> 'a -> 'a t = fun depth value ->
    if depth > 0 then
      Node (make (depth - 1) (value, value))
    else
      Leaf value

  let rec iter : 'a . ('a -> unit) -> 'a t -> unit = fun f tree ->
    match tree with
    | Leaf value -> f value
    | Node tree ->
        iter (fun (a, b) -> f a; f b) tree
end

File: tests/irregular/irregular.py
import ocaml
ocaml.add_dir("../../api/.ocaml_in_python_api.objs/byte/")
ocaml.add_dir(".irregular.objs/byte/")
ocaml.Dynlink.loadfile("irregular.cmxs")
from ocaml import Irregular
tree = Irregular.CompleteBinaryTree.make(3, "a")
accu = []
Irregular.CompleteBinaryTree.iter((lambda x: accu.append(x)), tree)
assert len(accu) == 8
tree = Irregular.CompleteBinaryTree.Node(Irregular.CompleteBinaryTree.Node(Irregular.CompleteBinaryTree.Leaf(((1, 2), (3, 4)))))
accu = []
Irregular.CompleteBinaryTree.iter((lambda x: accu.append(x)), tree)
assert accu == [1, 2, 3, 4]
